### DISCUSSION TITLE
Need a query function which returns an actual increase in counter values within a time interval and handles counter reset cases

### DISCUSSION THREAD

**Discussion by anantha96** [prometheus/prometheus#11466] [General]
Is there a functionality using which we can get the difference in the values of counter samples within the time interval, while handling counter reset case and excluding the extrapolate functionality? 

We tried using the increase() function, but it returns an extrapolated result. We have observed that the difference between the actual increase and the extrapolated result is considerably high.  

**Example:**
If we are calculating the increase in a metric "node_disk_read_bytes_total" every 5 mins, with prometheus scrape interval set as 1 min:

-  Consider the following sample values for the metric "node_disk_read_bytes_total" within a 5 mins interval :
            [23758450955264 **(F)**, 23758499419136, 23758518625280, 23758519292928, 23758519870464 **(L)**]

**Result of the increase function:**
- Extrapolated value returned by increase function over 5 mins: 86144000

**_Our requirement is a function which:_**
- _Handles the counters resets similar to increase function_
- _And returns the actual difference b/w the values of the first sample(F) and Last sample(L) within the specified interval : 68915200_

**Comment by natewaddoups** [comment]
According to this comment, what the OP is asking for is not possible with Grafana and Prometheus:

https://github.com/prometheus/prometheus/issues/3746#issuecomment-360826440

Or it wasn't possible in 2018. 

Is it possible today? I've been running into similar issues with my own Grafana charts.

**Comment by beorn7** [comment]
Short answer: there is no feature like this in PromQL, but we might add one.

I still doubt that people really know what they are wishing for if they say the would like "the actual difference b/w the values of the first sample(F) and Last sample(L) within the specified interval". E.g. if you say `increase(foo_total[100d])`, you might get just the difference over the last 5m, if the only samples in the 100d are from the last 5m.

I'm a bit more sold that people would like "increase between what the time series was now and 10m ago", so this would be something like `foo_total - foo_total offset 10m` with somehow added counter reset handling. I still assume most people don't really understand the implications of the selection of samples and their timing they are comparing here. I would still assume that most of the time people are better off with the extrapolated value PromQL provides.

_However_, people have created the most creative and most hilarious work-arounds to build themselves a much worse footgun, so I guess we need to provide this feature is some way.

**Comment by natewaddoups** [comment]
It's been a headache for my team since we switched from DataDog (where charts based on counters "just worked") to Prometheus+Grafana (where they are full of surprises). To be fair, I am not certain whether the root cause is Grafana or Prometheus, and I suspect there are quirks in both systems that make it hard to solve this problem:

<img width="833" alt="image" src="https://github.com/user-attachments/assets/63dff2bf-9819-42ff-91df-601f7e94e8c6" />

The problem was about fixed increase ranges, but I don't get consistent results using Grafana's dynamic ranges either.

Also from that thread:

"If you want exact results you need to use a logs-based monitoring system, not a metrics-based one."

I am using that workaround in places, but the charts take a long time to render. I really seems like it should be possible to produce accurate time-series charts with counters. But I'll confess that I don't have a clear picture of what the solution would be.
